saiph/reduction/utils/svd.py

```diff
@@ -1,26 +1,120 @@
 from sklearn.utils import extmath
 
 
-def SVD(
-    df: pd.DataFrame, svd_flip: bool = True
+def get_svd(
+    df: pd.DataFrame,
+    nf: Optional[int] = None,
+    svd_flip: bool = True,
+    seed: Optional[int] = None,
 ) -> Tuple[NDArray[Any], NDArray[Any], NDArray[Any]]:
     """Compute Singular Value Decomposition.
 
-    Parameters:
-        df: Matrix to decompose.
-        svd_flip: Whether to use svd_flip on U and V or not.
+    Arguments
+    ---------
+        df: Matrix to decompose, shape (m, n)
+        nf: target number of dimensions to retain (number of singular values). Default `None`
+        svd_flip: Whether to use svd_flip on U and V or not. Default `True`
+        seed: random seed. Default `None`
 
-    Returns:
-        U: Unitary matrix having left singular vectors as columns.
-        s: The singular values.
-        V: Unitary matrix having right singular vectors as rows.
+    Returns
+    -------
+        U: unitary matrix having left singular vectors as columns, shape (m,l)
+        S: vector of the singular values, shape (l,)
+        Vt: unitary matrix having right singular vectors as rows, shape (l,n)
     """
-    U, s, V = linalg.svd(df, full_matrices=False)
+    if nf is not None and nf != np.min(df.shape):
+        # Randomized SVD
+        U, S, Vt = get_direct_randomized_svd(df, q=2, l=nf, seed=seed)
+
+    else:
+        # Full SVD
+        U, S, Vt = linalg.svd(df, full_matrices=False)
+
     if svd_flip:
-        U, V = extmath.svd_flip(U, V)
-    return U, s, V
+        U, Vt = extmath.svd_flip(U, Vt)
+
+    return U, S, Vt
+
+
+def get_randomized_subspace_iteration(
+    A: NDArray[np.float_], l: int, q: int = 2, seed: Optional[int] = None
+) -> NDArray[np.float_]:
+    """Generate a subspace for more efficient SVD compuation using random methods.
+
+    From http://arxiv.org/abs/0909.4061, algorithm 4.4 page 27
+
+    Arguments
+    ---------
+        A: input matrix, shape (m, n)
+        l: target number of retained dimensions, l<min(m,n)
+        q: exponent of the power method. Higher this exponent, the more precise will be the SVD, but more complex to compute. Default `2`
+        seed: random seed. Default `None`
+
+    Returns
+    -------
+        Q: matrix whose range approximates the range of A, shape (m, l)
+    """
+    m, n = A.shape
+    np.random.seed(seed)
+    Omega = np.random.normal(loc=0, scale=1, size=(n, l))
+
+    # Initialization
+    Y = A @ Omega
+    Q, _ = np.linalg.qr(Y)
+
+    # Iteration
+    for i in range(q):
+        Ytilde = A.transpose() @ Q
+        Qtilde, _ = np.linalg.qr(Ytilde)
+        Y = A @ Qtilde
+        Q, _ = np.linalg.qr(Y)
+    return Q
+
+
+def get_direct_randomized_svd(
+    A: NDArray[np.float_], l: int, q: int = 2, seed: Optional[int] = None
+) -> Tuple[NDArray[Any], NDArray[Any], NDArray[Any]]:
+    """Compute a fixed-rank SVD approximation using random methods.
+
+    From http://arxiv.org/abs/0909.4061, algorithm 5.1 page 29
```

Comment:
```suggestion
    From https://arxiv.org/abs/0909.4061, algorithm 5.1 page 29
```
nit :smiley:

Reply:
Also, maybe add the name of the paper as well as the authors, so that if the link is dead, we can find it back again